nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/ConsumeMQTT.java

```diff
@@ -0,0 +1,341 @@
+            final MQTTQueueMessage mqttMessage = mqttQueue.peek();
+
+            Map<String, String> attrs = new HashMap<>();
+            attrs.put(BROKER_ATTRIBUTE_KEY, broker);
+            attrs.put(TOPIC_ATTRIBUTE_KEY, mqttMessage.getTopic());
+            attrs.put(QOS_ATTRIBUTE_KEY, String.valueOf(mqttMessage.getQos()));
+            attrs.put(IS_DUPLICATE_ATTRIBUTE_KEY, String.valueOf(mqttMessage.isDuplicate()));
+            attrs.put(IS_RETAINED_ATTRIBUTE_KEY, String.valueOf(mqttMessage.isRetained()));
+
+            messageFlowfile = session.putAllAttributes(messageFlowfile, attrs);
+
+            messageFlowfile = session.write(messageFlowfile, new OutputStreamCallback() {
+                @Override
+                public void process(final OutputStream out) throws IOException {
+                    out.write(mqttMessage.getPayload());
+                }
+            });
+
+            String transitUri = new StringBuilder(broker).append(mqttMessage.getTopic()).toString();
+            session.getProvenanceReporter().receive(messageFlowfile, transitUri);
+            session.transfer(messageFlowfile, REL_MESSAGE);
+            mqttQueue.remove(mqttMessage);
+            session.commit();
+        }
+    }
+
+    private class ConsumeMQTTCallback implements MqttCallback {
+
+        @Override
+        public void connectionLost(Throwable cause) {
+            logger.warn("Connection to " + broker + " lost", cause);
+            try {
+                reconnect();
+            } catch (MqttException e) {
+                logger.error("Connection to " + broker + " lost and callback re-connect failed.");
+            }
+        }
+
+        @Override
+        public void messageArrived(String topic, MqttMessage message) throws Exception {
+            logger.info("MQTT message arrived on topic:" + topic);
+            if (mqttQueue.size() >= maxQueueSize){
+                throw new IllegalStateException("The subscriber queue is full, cannot receive another message until the processor is scheduled to run.");
+            } else {
+                mqttQueue.add(new MQTTQueueMessage(topic, message));
+            }
+        }
+
+        @Override
+        public void deliveryComplete(IMqttDeliveryToken token) {
+            logger.warn("Received MQTT 'delivery complete' message to subscriber:"+ token);
+        }
+    }
+
+    // Public for testing
+    public void reconnect() throws MqttException {
+        synchronized (mqttClientConnectLock) {
+            if (!mqttClient.isConnected()) {
+                mqttClient = getMqttClient(broker, clientID, persistence);
+                mqttClient.setCallback(new ConsumeMQTTCallback());
+                mqttClient.connect(connOpts);
+                if(!subscribed){
+                    mqttClient.subscribe(topicFilter, qos);
+                    subscribed = true;
+                }
+            }
+        }
+    }
+
+    // Public for testing
+    public boolean isConnected(){
```

Comment:
So, Reflection is a very common pattern to be used for such things. Basically, unless you have any reason other then testing to make it public its fine, but if its only testing then consider creating a reflection-based callback in some test utils. 
Also, look at it from the different angle. The tighter the code base is the easier it will be to introduce changes to in the future. Making it public you are immediately adding it to any and all future discussions on breaking changes regardless if when/if they happen. 

Yes, I do see that you are not setting it to null explicitly and for that part of the comment its OK, yet it still feels a bit odd, but I am ok for now